<a href="https://colab.research.google.com/github/tanvisht/Data-Science-NYU-Stern/blob/Assignments/Challenge_1_Tanvish.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

PROBLEM 1

1.1 Load and inspect the data

In [2]:
from google.colab import files
uploaded = files.upload()

Saving student_performance.csv to student_performance.csv


In [4]:
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv('student_performance.csv')

# a) Display the dataset dimensions
print(f"Shape: {df.shape[0]} students and {df.shape[1]} columns")

# b) Inspect column names and data types
print("\n--- Column Types ---")
print(df.dtypes)

# c) Display the first 10 rows to verify data loading
print("\n--- First 10 Rows ---")
display(df.head(10))

# d) Check for missing values across all columns
print("\n--- Missing Values ---")
print(df.isnull().sum())

# e) Analyze the distribution of the target variable 'status'
print("\n--- Status Distribution ---")
# Check distribution of Dropout, Graduate, and Enrolled students
print(df['status'].value_counts())

Shape: 4424 students and 20 columns

--- Column Types ---
student_id                  int64
age                         int64
gender                     object
marital_status             object
course                     object
attendance                 object
admission_grade           float64
previous_grade            float64
scholarship                object
tuition_up_to_date         object
debtor                     object
semester1_enrolled          int64
semester1_approved          int64
semester1_grade           float64
semester2_enrolled          int64
semester2_approved          int64
semester2_grade           float64
semester1_success_rate    float64
semester2_success_rate    float64
status                     object
dtype: object

--- First 10 Rows ---


student_id  age  gender marital_status                course attendance  \
0           1   20    Male         Single    Animation & Design    Daytime   
1           2   19    Male         Single               Tourism    Daytime   
2           3   19    Male         Single  Communication Design    Daytime   
3           4   20  Female         Single            Journalism    Daytime   
4           5   45  Female        Married        Social Service    Evening   
5           6   50    Male        Married            Management    Evening   
6           7   18  Female         Single               Nursing    Daytime   
7           8   22    Male         Single               Tourism    Daytime   
8           9   21  Female         Single        Social Service    Daytime   
9          10   18  Female         Single        Social Service    Daytime   

   admission_grade  previous_grade scholarship tuition_up_to_date debtor  \
0            127.3           122.0          No                Yes     No   
1            142.5           160.0          No                 No     No   
2            124.8           122.0          No                 No     No   
3            119.6           122.0          No                Yes     No   
4            141.5           100.0          No                Yes     No   
5            114.8           133.1          No                Yes    Yes   
6            128.4           142.0         Yes                Yes     No   
7            113.1           119.0          No                 No     No   
8            129.3           137.0         Yes                Yes     No   
9            123.0           138.0          No                 No    Yes   

   semester1_enrolled  semester1_approved  semester1_grade  \
0                   0                   0              0.0   
1                   6                   6             14.0   
2                   6                   0              0.0   
3                   6                   6             13.4   
4                   6                   5             12.3   
5                   5                   5             11.9   
6                   7                   7             13.3   
7                   5                   0              0.0   
8                   6                   6             13.9   
9                   6                   5             11.4   

   semester2_enrolled  semester2_approved  semester2_grade  \
0                   0                   0              0.0   
1                   6                   6             13.7   
2                   6                   0              0.0   
3                   6                   5             12.4   
4                   6                   6             13.0   
5                   5                   5             11.5   
6                   8                   8             14.3   
7                   5                   0              0.0   
8                   6                   6             14.1   
9                   6                   2             13.5   

   semester1_success_rate  semester2_success_rate    status  
0                     0.0                     0.0   Dropout  
1                   100.0                   100.0  Graduate  
2                     0.0                     0.0   Dropout  
3                   100.0                    83.3  Graduate  
4                    83.3                   100.0  Graduate  
5                   100.0                   100.0  Graduate  
6                   100.0                   100.0  Graduate  
7                     0.0                     0.0   Dropout  
8                   100.0                   100.0  Graduate  
9                    83.3                    33.3   Dropout


--- Missing Values ---
student_id                0
age                       0
gender                    0
marital_status            0
course                    0
attendance                0
admission_grade           0
previous_grade            0
scholarship               0
tuition_up_to_date        0
debtor                    0
semester1_enrolled        0
semester1_approved        0
semester1_grade           0
semester2_enrolled        0
semester2_approved        0
semester2_grade           0
semester1_success_rate    0
semester2_success_rate    0
status                    0
dtype: int64

--- Status Distribution ---
status
Graduate    2209
Dropout     1421
Enrolled     794
Name: count, dtype: int64


1.2 Program Performance Dashboard

In [5]:
# 1.2 Program Performance Dashboard

# Define a custom aggregation function to count dropouts
def dropout_count(x):
    return (x == 'Dropout').sum()

# Aggregate data by Course to calculate key performance metrics
# Columns selected: Status, Admission Grade, Sem 1 Grade, Sem 2 Grade
program_stats = df.groupby('course').agg({
    'status': ['count', dropout_count],
    'admission_grade': 'mean',
    'semester1_grade': 'mean',
    'semester2_grade': 'mean'
})

# Flatten the MultiIndex columns and rename for clarity
program_stats.columns = [
    'Total Students',
    'Dropout Count',
    'Avg Admission Grade',
    'Avg Sem 1 Grade',
    'Avg Sem 2 Grade'
]

# Calculate the Dropout Rate percentage for each program
program_stats['Dropout Rate (%)'] = (program_stats['Dropout Count'] / program_stats['Total Students']) * 100

# Sort by Dropout Rate descending to identify high-risk programs
display(program_stats.sort_values(by='Dropout Rate (%)', ascending=False))

Total Students  Dropout Count  Avg Admission Grade  \
course                                                                        
Biofuel Production                   12              8           118.558333   
Equinculture                        141             78           132.407092   
Informatics Engineering             170             92           126.399412   
Basic Education                     192             85           123.162500   
Management                          648            270           125.894907   
Agronomy                            210             86           132.755714   
Oral Hygiene                         86             33           123.088372   
Animation & Design                  215             82           136.156279   
Tourism                             252             96           121.572222   
Marketing Management                268             95           121.462687   
Journalism                          331            101           127.958912   
Veterinary Nursing                  337             90           131.174184   
Social Service                      570            136           124.295439   
Communication Design                226             51           128.889381   
Nursing                             766            118           127.259138   

                         Avg Sem 1 Grade  Avg Sem 2 Grade  Dropout Rate (%)  
course                                                                       
Biofuel Production              9.650000         8.991667         66.666667  
Equinculture                   10.085816         9.685816         55.319149  
Informatics Engineering         8.564118         8.785882         54.117647  
Basic Education                11.469271        10.567188         44.270833  
Management                      9.699691         9.503086         41.666667  
Agronomy                       10.376190         9.476667         40.952381  
Oral Hygiene                   10.338372         9.768605         38.372093  
Animation & Design              2.068837         2.082791         38.139535  
Tourism                        10.462698         9.927381         38.095238  
Marketing Management           11.330224        10.682090         35.447761  
Journalism                     11.695770        11.036254         30.513595  
Veterinary Nursing             11.860534        11.161128         26.706231  
Social Service                 10.984386        10.618596         23.859649  
Communication Design           12.083628        11.196903         22.566372  
Nursing                        12.464491        12.355222         15.404700

1.2 Program Analysis

Highest & Lowest Dropout Rates: The program with the highest dropout rate is Biofuel Production (66.7%), followed closely by Equinculture (55.3%) and Informatics Engineering (54.1%). The program with the lowest dropout rate is Nursing (15.4%).

* Admission Grades vs. Retention: Programs with higher admission grades do not consistently have lower dropout rates. for example, Equinculture has a high average admission grade (approx. 132.4) yet maintains one of the highest dropout rates, suggesting that pre-admission academic performance is not the sole predictor of retention.

* Hypothesis: The high dropout rate in Informatics Engineering (54.1%) despite reasonable admission grades may be due to the rigorous nature of the curriculum or a "shock" in the first semester, as evidenced by their low average Semester 1 grade (8.56), which is the second lowest among all programs.

* Recommendation: For the Informatics Engineering program, I recommend implementing mandatory supplemental instruction or peer tutoring for first-year courses to improve Semester 1 performance, as early academic struggles appear to be a key driver of attrition in this major.

PROBLEM 2: Identifying Risk Factors


2.1 Academic Performance Patterns

In [10]:
# 2.1 Academic Performance Patterns

# --- Part A: Metrics by Status ---

# Calculate Semester 1 Success Rate (Approved / Enrolled)
# We use np.where to handle division by zero (if a student enrolled in 0 courses)
df['sem1_success_pct'] = np.where(df['semester1_enrolled'] > 0,
                                  df['semester1_approved'] / df['semester1_enrolled'],
                                  0)

# Identify students who failed ALL semester 1 courses (0 approved courses)
# This is a binary flag (1 = Failed All, 0 = Did not fail all)
df['failed_all_sem1'] = np.where(df['semester1_approved'] == 0, 1, 0)

# Group by Status to see patterns
# We calculate the mean to get averages and percentages
status_metrics = df.groupby('status').agg({
    'sem1_success_pct': 'mean',          # Average success rate
    'semester1_grade': 'mean',           # Average grade
    'failed_all_sem1': 'mean'            # % who failed all (mean of 0s and 1s = %)
})

print("--- Part A: Academic Patterns by Status ---")
display(status_metrics)


# --- Part B: At-Risk Analysis (< 50% Success) ---

# Define At-Risk: Success rate < 0.50 (50%)
df['at_risk'] = df['sem1_success_pct'] < 0.50

# 1. Percentage of all students who are at-risk
pct_at_risk = df['at_risk'].mean() * 100

# 2. Dropout rate among At-Risk students
# Filter for at-risk, then check dropout rate
at_risk_group = df[df['at_risk'] == True]
dropout_rate_risk = (at_risk_group['status'] == 'Dropout').mean() * 100

# 3. Dropout rate among NOT At-Risk students
safe_group = df[df['at_risk'] == False]
dropout_rate_safe = (safe_group['status'] == 'Dropout').mean() * 100

print("\n--- Part B: At-Risk Analysis ---")
print(f"Percentage of students considered 'At-Risk': {pct_at_risk:.2f}%")
print(f"Dropout Rate (At-Risk Group): {dropout_rate_risk:.2f}%")
print(f"Dropout Rate (Not At-Risk Group): {dropout_rate_safe:.2f}%")

--- Part A: Academic Patterns by Status ---


sem1_success_pct  semester1_grade  failed_all_sem1
status                                                      
Dropout           0.383994         7.256721         0.401126
Enrolled          0.688371        11.124559         0.089421
Graduate          0.903223        12.644047         0.034857


--- Part B: At-Risk Analysis ---
Percentage of students considered 'At-Risk': 23.24%
Dropout Rate (At-Risk Group): 77.43%
Dropout Rate (Not At-Risk Group): 18.40%


2.1 Academic Patterns & Risk Analysis

Predictive Power: Semester 1 performance is a definitive predictor of dropout. The data shows that students who dropout have a drastically lower success rate in their first semester compared to graduates.

Threshold Effectiveness: The 50% success rate threshold is highly effective. Students identified as "At-Risk" (passing fewer than half their classes) have a dropout rate of 77.4%, whereas students who pass more than half their classes have a dropout rate of only 18.4%. This clearly separates the high-risk population.

Timing of Intervention: Intervention must occur immediately after Semester 1. Waiting for Semester 2 is too late, as 40.1% of all dropouts failed every single course in their first semester, indicating they have already fully disengaged from the academic program.

Recommended Action: The university should automatically place any student with a Sem 1 success rate below 50% into a mandatory "Academic Probation & Recovery" program. This program would require bi-weekly meetings with an academic advisor to build a manageable course load for Semester 2.

Problem 2.2: Demographics and Financial Factors

In [8]:
# 2.2 Demographics and Financial Factors

# --- Part A: Impact of Scholarships ---
# Group by Scholarship (Yes/No)
scholarship_stats = df.groupby('scholarship').agg({
    'status': lambda x: (x == 'Dropout').mean() * 100, # Calculate Dropout Rate %
    'semester1_grade': 'mean',
    'semester2_grade': 'mean'
})
scholarship_stats.columns = ['Dropout Rate (%)', 'Avg Sem 1 Grade', 'Avg Sem 2 Grade']

print("--- Part A: Impact of Scholarships ---")
display(scholarship_stats)


# --- Part B: Impact of Age ---
# Create Age Bins: Traditional (17-22), Young Adult (23-30), Mature (31+)
age_bins = [0, 22, 30, 100]
age_labels = ['Traditional (17-22)', 'Young Adult (23-30)', 'Mature (31+)']
df['age_group'] = pd.cut(df['age'], bins=age_bins, labels=age_labels)

# Group by Age Group
# Note: observed=False handles categorical data warnings in newer pandas versions
age_stats = df.groupby('age_group', observed=False).agg({
    'status': ['count', lambda x: (x == 'Dropout').mean() * 100],
    'semester1_grade': 'mean'
})
age_stats.columns = ['Student Count', 'Dropout Rate (%)', 'Avg Sem 1 Grade']

print("\n--- Part B: Impact of Age ---")
display(age_stats)


# --- Part C: Financial Stability ---
# Define Financial Issues: Debtor ('Yes') OR Tuition Up to Date ('No')
# We use logical OR (|) to flag students with EITHER issue
df['financial_issues'] = (df['debtor'] == 'Yes') | (df['tuition_up_to_date'] == 'No')

financial_stats = df.groupby('financial_issues').agg({
    'status': lambda x: (x == 'Dropout').mean() * 100
})
financial_stats.columns = ['Dropout Rate (%)']
# Rename index for clarity
financial_stats.index = ['No Financial Issues', 'Has Financial Issues']

print("\n--- Part C: Financial Stability ---")
display(financial_stats)

--- Part A: Impact of Scholarships ---


Dropout Rate (%)  Avg Sem 1 Grade  Avg Sem 2 Grade
scholarship                                                    
No                  38.706767        10.172391         9.686977
Yes                 12.192903        12.058417        11.873430


--- Part B: Impact of Age ---


Student Count  Dropout Rate (%)  Avg Sem 1 Grade
age_group                                                            
Traditional (17-22)           3047         22.743682        11.216935
Young Adult (23-30)            715         52.167832         9.470210
Mature (31+)                   662         53.625378         9.254079


--- Part C: Financial Stability ---


Dropout Rate (%)
No Financial Issues          23.825227
Has Financial Issues         70.573248

2.2 Demographics and Financial Analysis:

Scholarships: Scholarships significantly improve retention. Students with scholarships have a dropout rate of 12.2%, which is roughly one-third of the rate for students without scholarships (38.7%). This suggests that financial support (and potentially the academic requirement to maintain it) is a strong protective factor.

Age Groups: The Mature (31+) age group has the highest dropout risk at 53.6%, followed closely by Young Adults at 52.2%. In contrast, Traditional students (17-22) perform much better with a dropout rate of only 22.7%, likely due to fewer external responsibilities such as full-time employment or family care.

Financial Status: Financial status is a critical determinant of success. Students flagged with financial issues (tuition debt or late payments) have a staggering dropout rate of 70.6%, compared to just 23.8% for those in good financial standing.

Recommendation: My #1 recommendation is to implement an Emergency Micro-Grant Program specifically targeting the students flagged with "Financial Issues." Since over 70% of this group drops out, resolving small tuition arrears or providing short-term financial relief could yield the highest immediate return on retention.

PROBLEM 3: Retention Intervention Strategy

To: University Administration

From: Tanvish Sameer Tuplondhe

Subject: Strategic Intervention Plan for Student Retention

Recommendation Memo

1. Target Audience: Based on the data analysis, intervention must target two specific high-risk groups: students with "Financial Issues" (tuition debt) and students with a Semester 1 success rate below 50%. These groups are prioritized because financial instability is the single strongest predictor of attrition, with a staggering 70.6% dropout rate for debtors compared to just 23.8% for financially stable students. Additionally, students who fail more than half of their first-semester courses show signs of immediate disengagement and are statistically unlikely to recover without support.
+1

2. Timing: Intervention should occur immediately following the release of Semester 1 grades. The data indicates that early academic performance is highly predictive; waiting until the end of the academic year is too late, as the majority of dropouts have already accumulated insurmountable grade deficits (average grade < 10) or ceased paying tuition by this point.
+1

3. Intervention Actions: I recommend a two-pronged approach: * Financial: Implement an Emergency Micro-Grant Program for students with tuition arrears, as data shows that financial aid (scholarships) reduces dropout rates by nearly 3x.

Academic: Enroll all students with <50% Semester 1 success in a mandatory "Academic Recovery Seminar" focused on study skills and peer tutoring.

Measurement: Success will be measured by tracking the retention rate of the "At-Risk" cohort entering Semester 2. The goal is to reduce the dropout rate among students with financial flags from 70.6% to 50% within the next academic year.